library/std/src/sync/once.rs

```diff
@@ -441,29 +447,29 @@ impl Once {
     }
 }
 
-fn wait(state_and_queue: &AtomicUsize, mut current_state: usize) {
+fn wait(state_and_queue: &AtomicPtr<Masked>, mut current_state: *mut Masked) {
     // Note: the following code was carefully written to avoid creating a
     // mutable reference to `node` that gets aliased.
     loop {
         // Don't queue this thread if the status is no longer running,
         // otherwise we will not be woken up.
-        if current_state & STATE_MASK != RUNNING {
+        if current_state.addr() & STATE_MASK != RUNNING {
             return;
         }
 
         // Create the node for our current thread.
         let node = Waiter {
             thread: Cell::new(Some(thread::current())),
             signaled: AtomicBool::new(false),
-            next: (current_state & !STATE_MASK) as *const Waiter,
+            next: current_state.with_addr(current_state.addr() & !STATE_MASK) as *const Waiter,
         };
-        let me = &node as *const Waiter as usize;
+        let me = &node as *const Waiter as *const Masked as *mut Masked;
 
         // Try to slide in the node at the head of the linked list, making sure
         // that another thread didn't just replace the head of the linked list.
         let exchange_result = state_and_queue.compare_exchange(
             current_state,
-            me | RUNNING,
+            me.with_addr(me.addr() | RUNNING),
```

Comment:
Yeah map_addr was added to the set of APIs we want to go with *after* I wrote all this so it's most only using with_addr, which isn't necessarily a *problem*.

Reply:
If we get really serious with this approach, I also like to imagine further conveniences, like abbreviating `p.with_addr(p.addr() | TAG)` as `p.tag_addr(TAG)` and similar for masking, which saves writing a `map_addr` closure.

Reply:
Agreed, let's be exuberantly maximalist about encapsulating these patterns. :)

Reply:
(My *really* maximalist thought is wondering how much raw pointer usage that's just for tagging could be replaced with completely safe wrappers like `Tagged<'a, T>` that consider how much alignment you get from `T` and handle masking on access automatically. But this can happen outside the standard library, I hope.)

Reply:
> (My really maximalist thought is wondering how much raw pointer usage that's just for tagging could be replaced with completely safe wrappers like Tagged<'a, T> that consider how much alignment you get from T and handle masking on access automatically. But this can happen outside the standard library, I hope.)

(Is this considered bad practice to link crates which happen to be yours on rust-lang discussion? Anyway, I have a crate that provides tagged pointer support, [ptr-union](https://lib.rs/crates/ptr-union), though it doesn't automatically detect guaranteed alignment yet, due to stable const generic limitations. I'm currently working on making it work under strict provenance / not use inttoptr.)